src/proxy/mod.rs

```diff
@@ -244,3 +246,64 @@ impl TryFrom<&str> for TraceParent {
         })
     }
 }
+pub fn get_original_src_from_fwded(req: &Request<Body>) -> Option<IpAddr> {
```

Comment:
We don't necessarily know its from another ztunnel, we just know its from someone that has a valid mTLS certificate. So its not "public internet" untrusted, but its also a source of risk.

Reply:
The traffic transmitted should be safe due mTLS certificate guaranteed(i.e. from ztunnel or waypoints) , am I right?
So the security  concern is 

1. at encode side, guarantee encoded IP is the trusted one? 
2. at decode side, guarantee decoded IP is also the tursted one?

Reply:
mTls doesn't encode which IP it's allowed to impersonate unless we also lookup the identity of the IP address. However, we may not know all info about all clients, so that doesn't work.

A header in hbone is essentially untrusted entirely. The only thing we know is they have an mTLS cert.

Reply:
sorry I think my comments got out of order since I forgot to hit "send"

Reply:
> 

Briefly went through the tweaked code,  we use HBONE tcp/ip layer to transfer to the original source info instead of http header. This could simplify and eliminate the security concern for forwarded head.

However spoofing the HBONE src_ip should be the same concern as non-meshed APP spoofing the src_ip?

Reply:
Yes but I think that is ok - we are preserving the semantics of Kubernetes here. If users chose to add source-based checks to their apps (which are not necessarily secure) they get the same behavior with and without ztunnel, making it transparent

Reply:
Not quite.

1. FORWARDED parsing is sensitive. We can make it safe though, so this is minor issue
2. FORWARDED fundamentally is unsafe because we are putting highly trusted data (source IP to spoof) into a low(er) trust payload (HBONE headers). This allows privilege escalation
3. A safer mechanism is to make ztunnel outbound just spoof the IP and then propogate the SRC IP from that rather than read headers; the diff I pasted above does that and I have tested it works.

Reply:
Regarding spoofing on the outbound - I don't believe this works for the general case, since it requires the inbound ztunnel to have a routing table configured to send traffic to spoofed IP via outbound ztunnel.

Reply:
It seems to work... not sure I understand exactly how though, the redirection/routing table configuration we have is pretty complex

Reply:
Some more thoughts on this:
1. Could it be the case that in tests workload IP == outbound ztunnel IP, so we're not really spoofing the IP address?
2. I don't believe it will work with waypoints, since the spoofed IP will be lost between waypoint and inbound ztunnel.

Reply:
...and some more:

3. The advantage of `Forwarded` header is that it works end-to-end without complex network routing setup.
4. If we're allowing peers to connect using spoofed IP address, then it seems that the vulnerability is already there.

Reply:
1. I agree we probably don't have adequate test coverage here. Its a bit tricky to begin with since CI happens to run as root (probably not a good idea... but that is how it is configured across the project today) but locally folks probably aren't.
2. is definitely true. In theory the waypoint could do it as well although I don't know that we want to give it NET_ADMIN. Generally the idea has been that ztunnel should be as transparent as possible while waypoint is essentially opting in to behavioral changes, so losing orig src with waypoint doesn't seem as bad (and in many cases could be sent with XFF header or similar?).

Reply:
2. I don't think that we should provide an inconsistent behavior of having/not having the original source depending on whether waypoint is used or not. Also, sending `X-Forwaded-For` (which is `Forwarded` without a source port) requires inbound ztunnel to parse IP from the header, so we're back at square one.
4. Agreed, except that the thing we're providing instead might not actually work ;)

Reply:
re re 2: XFF or Forwarded was meant for user application, I didn't mean ztunnel would do anything

4. Aside from waypoints (which, IMO, are a different story) is there case we know it doesn't work?

Reply:
re 4. I'm pretty sure it wouldn't work if workload IP != outbound ztunnel IP, but I haven't tested it.

I think one solution to do this properly and securely would be to map outbound ztunnel to a list of workloads it serves (either via xDS or maybe even in the TLS client certificate, since the outbound ztunnel or something mimicking it could technically spoof its source IP and pretend to be a different workload), so that we know that the connecting IP is allowed to forward traffic for the IP address from the `Forwarded` header.

Reply:
In our k8s test (https://github.com/istio/ztunnel/pull/264#discussion_r1047475451) it works if: there is no waypoint (as noted, this breaks the source), the server doesn't have a sidecar (this also breaks the source, unless TPROXY mode is enabled), and the src!=dst (self call; I think this is just an issue in our redirection logic that can be fixed).

The mapping was something I had considered but it seems concerning for an inbound ztunnel to need to know info about ALL possible clients. We don't always have global knowledge.

Reply:
What if we embed this information in the TLS certificate? Then we wouldn't need to have the global knowledge.

Reply:
That is an interesting idea... there have been a variety of other discussions that settled on "Put the IP in the cert" as well. My initial concern had been that we integrate with a variety of third party CAs which would all need to be updated to encode the IP (which may not be feasible for some), but it may be worth considering adding this requirement regardless given the possible benefits.

Reply:
Yeah, I think that's the right solution, especially if the IP address can be spoofed on the network...

But that's a much larger change, so in the interest of unblocking this PR, could we make this configurable for now?
1. non-transparent,
2. spoof outbound IP,
3. use `Forwarded` header (needed for waypoints and when IP address cannot be spoofed on the network).

Reply:
I am good with a short term approach for now. I started a doc on the options discussed here https://docs.google.com/document/d/17fDTMLg-22UPfh62d5wumAOIpbeoOyWix2WTE54hDNc/ so its not just in a github comment and asked some other folks to review. If the main concern is the waypoint hop I think we can handle it with a hybrid approach of using `Forwarded` in trusted cases but not all cases. I am not sure that handles all cases you are interested in, though.